mantaray/marshal.go

```diff
@@ -86,34 +182,100 @@ func (bb *bitsForBytes) iter(f func(byte) error) error {
 }
 
 // UnmarshalBinary deserialises a node
-func (n *Node) UnmarshalBinary(bytes []byte) error {
-	if len(bytes) < preambleSize {
+func (n *Node) UnmarshalBinary(data []byte) error {
+	if len(data) < nodeHeaderSize {
 		return ErrTooShort
 	}
-	n.entry = append([]byte{}, bytes[0:32]...)
+
+	n.obfuscationKey = append([]byte{}, data[0:nodeObfuscationKeySize]...)
+
+	// perform XOR decryption on bytes after obfuscation key
+	xorDecryptedBytes := make([]byte, len(data))
+
+	copy(xorDecryptedBytes, data[0:nodeObfuscationKeySize])
+
+	for i := nodeObfuscationKeySize; i < len(data); i += nodeObfuscationKeySize {
+		end := i + nodeObfuscationKeySize
+		if end > len(data) {
+			end = len(data)
+		}
+
+		decrypted := encryptDecrypt(data[i:end], n.obfuscationKey)
+		copy(xorDecryptedBytes[i:end], decrypted)
+	}
+
+	data = xorDecryptedBytes
+
+	// Verify version hash.
+	if versionHash := append([]byte{}, data[nodeObfuscationKeySize:nodeObfuscationKeySize+versionHashSize]...); !bytes.Equal(versionHash, version01HashBytes) {
+		return fmt.Errorf("invalid version hash %x", versionHash)
+	}
+
+	refBytesSize := int(data[nodeHeaderSize-1])
+
+	n.entry = append([]byte{}, data[nodeHeaderSize:nodeHeaderSize+refBytesSize]...)
+	offset := nodeHeaderSize + refBytesSize // skip entry
 	n.forks = make(map[byte]*fork)
-	offset := preambleSize
 	bb := &bitsForBytes{}
-	bb.fromBytes(bytes[32:])
-	bb.iter(func(b byte) error {
+	bb.fromBytes(data[offset:])
+	offset += 32 // skip forks
+	err := bb.iter(func(b byte) error {
 		f := &fork{}
-		f.fromBytes(bytes[offset : offset+forkSize])
+
+		if len(data) < offset+nodeForkPreReferenceSize+refBytesSize {
+			err := fmt.Errorf("not enough bytes for node fork: %d (%d)", (len(data) - offset), (nodeForkPreReferenceSize + refBytesSize))
+			return fmt.Errorf("%w on byte '%x'", err, []byte{b})
+		}
+
+		err := f.fromBytes(data[offset : offset+nodeForkPreReferenceSize+refBytesSize])
+		if err != nil {
+			return fmt.Errorf("%w on byte '%x'", err, []byte{b})
+		}
+
 		n.forks[b] = f
-		offset += forkSize
+		offset += nodeForkPreReferenceSize + refBytesSize
```

Comment:
lets keep `forkSize`  and define like this

Reply:
The `forkSize` variable was removed because it is fixed value. Current implementation supports different reference sizes, non-encrypted and encrypted.

The reference size is read from header and used during unmarshal process.